lib/Dialect/Calyx/Transforms/CompileControl.cpp

```diff
@@ -0,0 +1,190 @@
+    auto source = groupOp.getDoneOp().src();
+    auto doneOpValue =
+        !guard ? source
+               : builder.create<comb::AndOp>(wires->getLoc(), guard, source);
+
+    // Build the Guard for the `go` signal of the current group being walked.
+    // The group should begin when:
+    // (1) the current step in the fsm is reached, and
+    // (2) the done signal of this group is not high.
+    auto eqCmp = builder.create<comb::ICmpOp>(
+        wires->getLoc(), comb::ICmpPredicate::eq, fsmOut, fsmCurrentState);
+    auto notDone =
+        builder.create<comb::XorOp>(wires->getLoc(), doneOpValue, oneConstant);
+    auto groupGoGuard =
+        builder.create<comb::AndOp>(wires->getLoc(), eqCmp, notDone);
+
+    // Guard for the `in` and `write_en` signal of the fsm register. These are
+    // driven when the group has completed.
+    builder.setInsertionPoint(seqGroup);
+    auto groupDoneGuard =
+        builder.create<comb::AndOp>(wires->getLoc(), eqCmp, doneOpValue);
+
+    // Directly update the GroupGoOp of the current group being walked.
+    auto goOp = groupOp.getGoOp();
+    assert(goOp && "The Go Insertion pass should be run before this.");
+    goOp->setOperand(0, oneConstant);
+    goOp->insertOperands(1, {groupGoGuard});
+
+    // Add guarded assignments to the fsm register `in` and `write_en` ports.
+    fsmNextState =
+        createConstant(builder, wires->getLoc(), fsmBitWidth, fsmIndex + 1);
+    builder.setInsertionPointToEnd(seqGroupBody);
+    builder.create<AssignOp>(wires->getLoc(), fsmIn, fsmNextState,
+                             groupDoneGuard);
+    builder.create<AssignOp>(wires->getLoc(), fsmWriteEn, oneConstant,
+                             groupDoneGuard);
+    // Increment the fsm index for the next group.
+    ++fsmIndex;
+  });
+
+  // Build the final guard for the new Seq group's GroupDoneOp. This is defined
+  // by the fsm's final state.
+  builder.setInsertionPoint(seqGroup);
+  auto isFinalState = builder.create<comb::ICmpOp>(
+      wires->getLoc(), comb::ICmpPredicate::eq, fsmOut, fsmNextState);
+
+  // Insert the respective GroupDoneOp.
+  builder.setInsertionPointToEnd(seqGroupBody);
+  builder.create<GroupDoneOp>(seqGroup->getLoc(), oneConstant, isFinalState);
+
+  // Add continuous wires to reset the `in` and `write_en` ports of the fsm
+  // when the SeqGroup is finished executing.
+  builder.setInsertionPointToEnd(wiresBody);
+  auto zeroConstant = createConstant(builder, wires->getLoc(), fsmBitWidth, 0);
+  builder.create<AssignOp>(wires->getLoc(), fsmIn, zeroConstant, isFinalState);
+  builder.create<AssignOp>(wires->getLoc(), fsmWriteEn, oneConstant,
+                           isFinalState);
+
+  // Replace the SeqOp with an EnableOp.
+  builder.setInsertionPoint(seq);
+  auto newEnableOp =
+      builder.create<EnableOp>(seq->getLoc(), seqGroup.sym_name());
+  seq->erase();
+
+  // Keep a list of compiled groups associated with the new EnableOp.
+  newEnableOp->setAttr("groups",
```

Comment:
Just curious- are these going to be used elsewhere? I didn't see this attribute in the ODS for `SeqOp`, so maybe define it as an `OptionalAttr` to make it explicit (and I think you will get a nice setter method generated).

Reply:
Ok. This is mostly for debugging purposes, and so that we maintain the invariant that each group in `wires` is referenced somewhere in `control` throughout the lowering process. I think OptionalAttr is a good idea.